Problem Sets/extras.rkt

```diff
@@ -90,39 +90,66 @@
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-;; check-location : String String -> Void
-;; GIVEN: a 2 digit problem set number NN and the name of the file
-;; being qualified.
-;; EFFECT: throws an error if this file is not a directory of the form
-;; pdp-*./setNN 
-(define (check-location NN correct-file-name)
-  (define path-elements (explode-path (current-directory)))
+;; (check-location NN correct-file-name) : module-level form
+(define-syntax (check-location stx)
+  (syntax-case stx ()
+    [(_ NN correct-file-name)
+     (case (syntax-local-context)
+       [(module)
+        #`(check-location-actual (#%expression NN)
+                                 (#%expression correct-file-name)
+                                 (quote #,(syntax-source stx)))
+        ]
+       [(top-level)
+        (raise-syntax-error #f
+          "`check-location' may only be used inside a module"
+          stx)]
+       [else
+        (raise-syntax-error #f
+          "found a use of `check-location' that is not at the top level"
+          stx)])]))
+
+;; check-location-actual : String String Path -> Void
+;; GIVEN: a 2 digit problem set number NN, the expected name of the file
+;; being qualified, and the actual path to the file
+;; EFFECT: throws an error if this file is not in a directory of the form
+;; pdp-*./setNN or does not have the correct name.
+(define (check-location-actual NN correct-file-name actual-file-path)
+  (define path-elements (explode-path actual-file-path))
   (define path-len (length path-elements))
   (define correct-folder-name (string-append "set" NN))
   (cond
-    [(>= path-len 2) 
-     (define set-folder (path->string (last path-elements)))
-     (define pdp-folder (path->string (list-ref path-elements (- path-len 2))))
+    [(>= path-len 3)
+     (define actual-file-name (path->string (last path-elements)))
```

Comment:
Consider `(list-ref path-elements (- path-len 1))` for symmetry with the others instead of `(last path-elements)`.